sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -70,437 +75,362 @@ object AggregateUtils {
-            resultCb = if (resultCvs.isEmpty) {
-              new ColumnarBatch(Seq().toArray, 0)
-            } else {
-              numOutputRows += resultCvs.head.getBase.getRowCount
-              new ColumnarBatch(resultCvs.toArray, resultCvs.head.getBase.getRowCount.toInt)
-            }
-            numOutputBatches += 1
-            success = true
-            new Iterator[ColumnarBatch] {
-              TaskContext.get().addTaskCompletionListener[Unit] { _ =>
-                if (resultCb != null) {
-                  resultCb.close()
-                  resultCb = null
-                }
-              }
+    val shims = ShimLoader.getSparkShims
+    val ordering = groupingExpressions.map(shims.sortOrder(_, Ascending, NullsFirst))
+    val groupingAttributes = groupingExpressions.map(_.asInstanceOf[NamedExpression].toAttribute)
+    val aggBufferAttributes = groupingAttributes ++
+        aggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes)
+    val sorter = new GpuSorter(ordering, aggBufferAttributes)
+    val aggregates = boundExpressions.aggModeCudfAggregates.flatMap(_._2)
+    val aggBatchTypes = groupingExpressions.map(_.dataType) ++ aggregates.map(_.dataType)
+
+    // Use the out of core sort iterator to sort the batches by grouping key
+    outOfCoreIter = Some(GpuOutOfCoreSortIterator(
+      aggregatedBatchIter,
+      sorter,
+      LazilyGeneratedOrdering.forSchema(TrampolineUtil.fromAttributes(groupingAttributes)),
+      targetBatchSize,
+      totalTime = NoopMetric,
+      sortTime = metrics.sortTime,
+      outputBatches = NoopMetric,
+      outputRows = NoopMetric,
+      peakDevMemory = NoopMetric,
+      spillCallback = metrics.spillCallback))
+
+    // The out of core sort iterator does not guarantee that a batch contains all of the values
+    // for a particular key, so add a key batching iterator to enforce this. That allows each batch
+    // to be merge-aggregated safely since all values associated with a particular key are
+    // guaranteed to be in the same batch.
+    val keyBatchingIter = new GpuKeyBatchingIterator(
+      outOfCoreIter.get,
+      sorter,
+      aggBatchTypes.toArray,
+      targetBatchSize,
+      numInputRows = NoopMetric,
+      numInputBatches = NoopMetric,
+      numOutputRows = NoopMetric,
+      numOutputBatches = NoopMetric,
+      collectTime = NoopMetric,
+      concatTime = metrics.concatTime,
+      totalTime = NoopMetric,
+      peakDevMemory = NoopMetric,
+      spillCallback = metrics.spillCallback)
+
+    // Finally wrap the key batching iterator with a merge aggregation on the output batches.
+    new Iterator[ColumnarBatch] {
+      override def hasNext: Boolean = keyBatchingIter.hasNext
+
+      override def next(): ColumnarBatch = {
+        // batches coming out of the sort need to be merged
+        withResource(keyBatchingIter.next()) { unmergedBatch =>
+          computeAggregate(GpuColumnVector.extractColumns(unmergedBatch), merge = true)
```

Comment:
It might speed things up a lot if we can tell `computeAggregte` that the data is now sorted by the grouping keys. This is especially true for aggregations that require the data to be sorted.